src/main/java/org/opensearch/flowframework/workflow/AbstractRetryableWorkflowStep.java

```diff
@@ -8,27 +8,133 @@
-
+    private static final Logger logger = LogManager.getLogger(AbstractRetryableWorkflowStep.class);
     /** The maximum number of transport request retries */
     protected volatile Integer maxRetry;
+    private final MachineLearningNodeClient mlClient;
+    private final FlowFrameworkIndicesHandler flowFrameworkIndicesHandler;
 
     /**
      * Instantiates a new Retryable workflow step
      * @param settings Environment settings
      * @param clusterService the cluster service
+     * @param mlClient machine learning client
+     * @param flowFrameworkIndicesHandler FlowFrameworkIndicesHandler class to update system indices
      */
-    public AbstractRetryableWorkflowStep(Settings settings, ClusterService clusterService) {
+    public AbstractRetryableWorkflowStep(
+        Settings settings,
+        ClusterService clusterService,
+        MachineLearningNodeClient mlClient,
+        FlowFrameworkIndicesHandler flowFrameworkIndicesHandler
+    ) {
         this.maxRetry = MAX_GET_TASK_REQUEST_RETRY.get(settings);
         clusterService.getClusterSettings().addSettingsUpdateConsumer(MAX_GET_TASK_REQUEST_RETRY, it -> maxRetry = it);
+        this.mlClient = mlClient;
+        this.flowFrameworkIndicesHandler = flowFrameworkIndicesHandler;
+    }
+
+    /**
+     * Retryable get ml task
+     * @param workflowId the workflow id
+     * @param nodeId the workflow node id
+     * @param future the workflow step future
+     * @param taskId the ml task id
+     * @param retries the current number of request retries
+     * @param workflowStep the workflow step which requires a retry get ml task functionality
+     */
+    void retryableGetMlTask(
+        String workflowId,
+        String nodeId,
+        CompletableFuture<WorkflowData> future,
+        String taskId,
+        int retries,
+        String workflowStep
+    ) {
+        mlClient.getTask(taskId, ActionListener.wrap(response -> {
+            MLTaskState currentState = response.getState();
+            if (currentState != MLTaskState.COMPLETED) {
+                if (Stream.of(MLTaskState.FAILED, MLTaskState.COMPLETED_WITH_ERROR).anyMatch(x -> x == currentState)) {
+                    // Model registration failed or completed with errors
+                    String errorMessage = workflowStep + " failed with error : " + response.getError();
+                    logger.error(errorMessage);
+                    future.completeExceptionally(new FlowFrameworkException(errorMessage, RestStatus.BAD_REQUEST));
+                } else {
+                    // Task still in progress, attempt retry
+                    throw new IllegalStateException(workflowStep + " is not yet completed");
+                }
+            } else {
+                try {
+                    logger.info(workflowStep + " successful for {} and modelId {}", workflowId, response.getModelId());
+                    String resourceName = WorkflowResources.getResourceByWorkflowStep(getName());
```

Comment:
would the getName() here return the step it was called from or a `get_task` step

Reply:
Yes. From the enum we have in `WorkflowResources`.